src/python/CRABClient/JobType/Analysis.py

```diff
@@ -60,14 +60,23 @@ def run(self, filecacheurl = None):
             configArguments['inputdata'] = self.config.Data.inputDataset
 #        configArguments['ProcessingVersion'] = getattr(self.config.Data, 'processingVersion', None)
 
-        # Create CMSSW config
+        ## Create CMSSW config.
         self.logger.debug("self.config: %s" % self.config)
         self.logger.debug("self.config.JobType.psetName: %s" % self.config.JobType.psetName)
-        cmsswCfg = CMSSWConfig(config=self.config, logger=self.logger,
-                               userConfig=self.config.JobType.psetName)
+        ## The loading of a CMSSW pset in the CMSSWConfig constructor is not idempotent
+        ## in the sense that a second loading of the same pset may not produce the same
+        ## result. Therefore there is a cache in CMSSWConfig to avoid loading any CMSSW
+        ## pset twice. However, some "complicated" psets seem to evade the caching.
+        ## Thus, to be safe, keep the CMSSWConfig instance in a class variable, so that
+        ## it can be reused later if wanted (for example, in PrivateMC when checking if
+        ## the pset has an LHE source) instead of having to load the pset again.
+        ## As for what does "complicated" psets mean, Daniel Riley said that there are
+        ## some psets where one module modifies the configuration from another module.
+        self.cmsswCfg = CMSSWConfig(config=self.config, logger=self.logger,
+                                    userConfig=self.config.JobType.psetName)
```

Comment:
Andres takes seriously the good old saying: "Always code as if the guy who
ends up maintaining your code will be a violent psychopath who knows where
you live."
Il 17/apr/2015 19:44 "Eric Vaandering" <notifications@github.com> ha
scritto:

> In src/python/CRABClient/JobType/Analysis.py
> <https://github.com/dmwm/CRABClient/pull/4452#discussion_r28614621>:
>
> >          self.logger.debug("self.config: %s" % self.config)
> >          self.logger.debug("self.config.JobType.psetName: %s" % self.config.JobType.psetName)
> > -        cmsswCfg = CMSSWConfig(config=self.config, logger=self.logger,
> > -                               userConfig=self.config.JobType.psetName)
> > +        ## The loading of a CMSSW pset in the CMSSWConfig constructor is not idempotent
> > +        ## in the sense that a second loading of the same pset may not produce the same
> > +        ## result. Therefore there is a cache in CMSSWConfig to avoid loading any CMSSW
> > +        ## pset twice. However, some "complicated" psets seem to evade the caching.
> > +        ## Thus, to be safe, keep the CMSSWConfig instance in a class variable, so that
> > +        ## it can be reused later if wanted (for example, in PrivateMC when checking if
> > +        ## the pset has an LHE source) instead of having to load the pset again.
> > +        ## As for what does "complicated" psets mean, Daniel Riley said that there are
> > +        ## some psets where one module modifies the configuration from another module.
> > +        self.cmsswCfg = CMSSWConfig(config=self.config, logger=self.logger,
> > +                                    userConfig=self.config.JobType.psetName)
>
> No on can fault you for not documenting this :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dmwm/CRABClient/pull/4452/files#r28614621>.
>